internal/rpc/core/mempool.go

```diff
@@ -60,87 +60,57 @@ func (env *Environment) BroadcastTxSync(ctx *rpctypes.Context, tx types.Tx) (*co
-		return nil, fmt.Errorf("max_subscriptions_per_client %d reached", env.Config.MaxSubscriptionsPerClient)
-	}
-
-	// Subscribe to tx being committed in block.
-	subCtx, cancel := context.WithTimeout(ctx.Context(), SubscribeTimeout)
-	defer cancel()
-	q := types.EventQueryTxFor(tx)
-	deliverTxSub, err := env.EventBus.Subscribe(subCtx, subscriber, q)
-	if err != nil {
-		err = fmt.Errorf("failed to subscribe to tx: %w", err)
-		env.Logger.Error("Error on broadcast_tx_commit", "err", err)
-		return nil, err
-	}
-	defer func() {
-		args := tmpubsub.UnsubscribeArgs{Subscriber: subscriber, Query: q}
-		if err := env.EventBus.Unsubscribe(context.Background(), args); err != nil {
-			env.Logger.Error("Error unsubscribing from eventBus", "err", err)
-		}
-	}()
-
-	// Broadcast tx and wait for CheckTx result
-	checkTxResCh := make(chan *abci.Response, 1)
-	err = env.Mempool.CheckTx(
+	resCh := make(chan *abci.Response, 1)
+	err := env.Mempool.CheckTx(
 		ctx.Context(),
 		tx,
-		func(res *abci.Response) { checkTxResCh <- res },
+		func(res *abci.Response) { resCh <- res },
 		mempool.TxInfo{},
 	)
 	if err != nil {
-		env.Logger.Error("Error on broadcastTxCommit", "err", err)
-		return nil, fmt.Errorf("error on broadcastTxCommit: %v", err)
+		return nil, err
 	}
 
-	checkTxResMsg := <-checkTxResCh
-	checkTxRes := checkTxResMsg.GetCheckTx()
+	r := (<-resCh).GetCheckTx()
 
-	if checkTxRes.Code != abci.CodeTypeOK {
+	if !indexer.KVSinkEnabled(env.EventSinks) {
 		return &coretypes.ResultBroadcastTxCommit{
-			CheckTx:   *checkTxRes,
-			DeliverTx: abci.ResponseDeliverTx{},
-			Hash:      tx.Hash(),
-		}, nil
+				CheckTx: *r,
+				Hash:    tx.Hash(),
+			},
+			errors.New("cannot wait for commit because kvEventSync is not enabled")
 	}
 
-	// Wait for the tx to be included in a block or timeout.
-	select {
-	case msg := <-deliverTxSub.Out(): // The tx was included in a block.
-		deliverTxRes := msg.Data().(types.EventDataTx)
-		return &coretypes.ResultBroadcastTxCommit{
-			CheckTx:   *checkTxRes,
-			DeliverTx: deliverTxRes.Result,
-			Hash:      tx.Hash(),
-			Height:    deliverTxRes.Height,
-		}, nil
-	case <-deliverTxSub.Canceled():
-		var reason string
-		if deliverTxSub.Err() == nil {
-			reason = "Tendermint exited"
-		} else {
-			reason = deliverTxSub.Err().Error()
+	startAt := time.Now()
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Context().Done():
+			env.Logger.Error("Error on broadcastTxCommit",
+				"duration", time.Since(startAt),
+				"err", err)
+			return &coretypes.ResultBroadcastTxCommit{
+					CheckTx: *r,
+					Hash:    tx.Hash(),
+				}, fmt.Errorf("timeout waiting for commit of tx %s (%s)",
+					tx.String(), time.Since(startAt))
+		case <-timer.C:
+			txres, err := env.Tx(ctx, tx.Hash(), false)
+			if err != nil {
+				jitter := 100*time.Millisecond + time.Duration(rand.Int63n(int64(time.Second))) // nolint: gosec
```

Comment:
Possibly a good case for a backoff? E.g., https://play.golang.org/p/juEn6ZbIW8H

Reply:
the one wrinkle is that we don't actually have that much time to work with because the context gets canceled pretty quickly.

Reply:
Sure, I just figured it would be a little easier to read. I had to read this expression multiple times to convince myself we're adding a random fraction of the nanoseconds in a second here. 🙂  Even if we don't actually want to back off, or not back off that much, we could set the multiplier to 1.

Reply:
I added a 100ms backoff per attempt, plus a random amount of time between 0.1s and 1.1s... I suspect the details don't matter a lot, and like a consistent 250ms would also probably be *fine*.